src/Orleans.Transactions/DistributedTM/TransactionAgent.cs

```diff
@@ -1,271 +1,421 @@
+            logger.Trace($"abort {transactionInfo} {string.Join(",", participants.Select(p => p.ToString()))}");
+
+        // send one-way abort messages to release the locks and roll back any updates
+        await Task.WhenAll(participants.Select(p => p.Reference.AsReference<ITransactionalResourceExtension>()
+             .Abort(p.Name, transactionInfo.TransactionId)));
+    }
+
+    public async Task Transaction(TransactionOption transactionOption, Func<Task> transactionScope)
+    {
+        if (transactionScope is null)
+        {
+            throw new ArgumentNullException(nameof(transactionScope));
+        }
+
+        // Pick up ambient transaction scope
+        var ambientTransactionInfo = TransactionContext.GetTransactionInfo();
+
+        if (ambientTransactionInfo is not null && transactionOption == TransactionOption.Suppress)
+        {
+            throw new NotSupportedException("Scope cannot be executed within a transaction.");
+        }
+
+        if (ambientTransactionInfo is null && transactionOption == TransactionOption.Join)
+        {
+            throw new NotSupportedException("Scope cannot be executed outside of a transaction.");
+        }
+
+        try
+        {
+            switch (transactionOption)
+            {
+                case TransactionOption.Create:
+                    await RunTransaction(null, transactionScope);
+                    break;
+                case TransactionOption.Join:
+                    await RunTransaction(ambientTransactionInfo, transactionScope);
+                    break;
+                case TransactionOption.CreateOrJoin:
+                    await RunTransaction(ambientTransactionInfo, transactionScope);
+                    break;
+                case TransactionOption.Suppress:
+                    await RunSupressedTransaction(transactionScope);
+                    break;
+                case TransactionOption.Supported:
+                    await RunSupportedTransaction(transactionScope);
+                    break;
+                case TransactionOption.NotAllowed:
+                    await RunDisallowedTransaction(ambientTransactionInfo, transactionScope);
+                    break;
+                default:
+                    throw new NotImplementedException();
             }
+        }
+        finally
+        {
+            // Restore ambient transaction context, if any
+            TransactionContext.SetTransactionInfo(ambientTransactionInfo);
+        }
+    }
 
-            if (logger.IsEnabled(LogLevel.Trace))
-                logger.Trace($"{stopwatch.Elapsed.TotalMilliseconds:f2} finish {transactionInfo.TransactionId}");
+    private static async Task RunDisallowedTransaction(TransactionInfo ambientTransactionInfo, Func<Task> transactionScope)
+    {
+        if (ambientTransactionInfo is not null)
+        {   // No transaction is allowed within scope
+            throw new InvalidOperationException();
```

Comment:
A descriptive error message would be appropriate here

Reply:
Thanks for feedback, I've included an error message